### ISSUE TITLE
Rethink the missing/default field attributes

### ISSUE THREAD

**Issue by touilleMan** [Scille/umongo#36]
Those attributes come from marshmallow, which is all about serialize/deserialize a document:
- `missing` is used when the field is not present in the dict to deserialize
- `default` is used when the field is not present in the document to serialize

However in umongo the focus shifts toward the document itself, so those terms seems a bit clunky
- the `missing` object is already used inside umongo's `DataProxy` to represent field not present in MongoDB. Hence the `missing` attribute should mean what value to return if the value is `missing`
- the `default` attribute sounds like the value to set by default to the field when creating a new document

In a nutshell the meaning of `missing` and `default` are reversed between mashmallow and umongo...

What I'm thinking to do:
- Remove the `missing` attribute (in fact just hide it from `abstract.BaseField` constructor and documentation)
- Only use the `default` attribute for both missing and default (in marshmallow's logic)
- Add methods to get from the umongo document an equivalent pure mashmallow Schema (see #34 )

The idea is to hide mashmallow logic to expose a more consistent API from the umongo user's point of view. Then provide a way to get back a pure Marshmallow Schema when needed to do all the custom.

example:

``` python
@instance.register
class Person(Document):
    name = fields.StrField(default='John Doe')

p = Person()
# Default is set
assert p.name == 'John Doe'
# Default value will be written in database
assert p._data.get('name') == 'John Doe'
# If we want more cunning behavior (e.g. only save in database non-default value)
# we should use a `MethodField` to define custom method for serialization/deseriazation

del p.name
assert p._data.get('name') == missing
# If not present on database, we switch back to default value as well
assert p.name == 'John Doe'

# Now it's customization time !
PersonSchema = p.get_marshmallow_schema()
class MyCustomSchema(PersonSchema):
    ...

# It could be also useful to provide method to get a specific field
class MyCustomSchema2(marshmallow.Schema)
    name = p.get_marshmallow_field('name')
```

@lafrech What do you think ?

**Comment by lafrech** [comment]
Honestly, I don't really grasp the implications on the usability (use cases) and the internals of umongo.

On first thought, fitting two different attributes/concepts into one sounds like losing information/possibilities, but I don't have a use case for a `default` value that would be different from a `missing` value, so if exposing only `default` is possible, then it would make things look simpler to the user.

```py
del p.name
assert p._data.get('name') == missing
# If not present on database, we switch back to default value as well
assert p.name == 'John Doe'
```

So once `name` is deleted, it will never appear in DB unless set explicitly?

This introduces a difference between a field set by default at document creation (value written in DB) and a field set by default at document load from DB (value missing in DB). Sounds strange but not necessarily an issue. I can't tell.

We may want to think about the case of an application update that would change the default value. The consequences would be different if the default value is always written in DB or left missing there and only viewed when using the document.

(I'm not sure I'm making much sense. It's a bit late, again. BTW, I'm on holidays for 3 weeks and my graphic card just let me down, so don't expect too much reactivity, but I'll try to keep an eye on my inbox...)

**Comment by touilleMan** [comment]
> This introduces a difference between a field set by default at document creation (value written in DB) and a field set by default at document load from DB (value missing in DB).

You're right on this one, this brings inconsistency which should be avoided.
So I wolud say `del doc.name` should be equivalent to the document initialization with no value for this field.

Here is the updated example
```python
@instance.register
class Person(Document):
    name = fields.StrField(default='John Doe')

p = Person()
# Default is set
assert p.name == 'John Doe'
# Default value will be written in database
assert p._data.get('name') == 'John Doe'
# If we want more cunning behavior (e.g. only save in database non-default value)
# we should use a `MethodField` to define custom method for serialization/deseriazation

# Set field to a custom value
p.name = 'Derp'
assert p._data.get('name') == 'Derp'

# Clearing the field will revert it to it default value
del p.name
assert p._data.get('name') == 'John Doe'

# Now it's customization time !
PersonSchema = p.get_marshmallow_schema()
class MyCustomSchema(PersonSchema):
    ...

# It could be also useful to provide method to get a specific field
class MyCustomSchema2(marshmallow.Schema)
    name = p.get_marshmallow_field('name')
```

PS: With the new GeForce out, your dead GPU sounds like the perfect excuse to your wife to get one !

**Comment by lafrech** [comment]
>    # Clearing the field will revert it to it default value
    del p.name
    assert p._data.get('name') == 'John Doe'

This may look unintuitive, but it seems fine to me. After all, when a default value exists on a field, there is no way to create a document without this attribute, as it will be created with the default value. The only way to not set a value at document creation is in fact to explicitly set it to a no-value (`None`). Likewise, if the user wants to remove a value on an existing document, he should set it to `None` rather than using `del`. Sounds coherent.

Your updated proposal looks fine to me.

The way I understand it, in the schema returned by `get_marshmallow_schema()`, the fields should have neither `missing` nor `default` as those values are not related to the `default` attribute in umongo. The user would customize those after the schema export.

**Comment by lafrech** [comment]
For the record, I just found out it is currently possible to create a field with a default value that is invalid.

```py

@instance.register
class User(Document):
    name = StringField(
        default='Eric',
        validate=validate.OneOf(('Stan', 'Kyle', 'Kenny'))
    )

user = User()
print(user.name)
# Eric
print(user._data)
# <UserDataProxy({'id': <marshmallow.missing>, 'name': <marshmallow.missing>})>
print("Committing with name = {}".format(user.name))
user.commit(io_validate_all=True)
# No error

user.name = 'Eric'
# An error is raised
```

When calling `user=User()`, the value for field name is `missing` and therefore passes validation, but  `user.name` returns `Eric` which is not valid.

I think your proposal fixes this:

```py
p = Person()
# Default is set
assert p.name == 'John Doe'
# Default value will be written in database
assert p._data.get('name') == 'John Doe'
```

**Comment by lafrech** [comment]
There might be an issue currently with passing default field values in `as_marshmallow_field`. We just fell into this:

```python
@instance.register
class Whatever(Document):
    date = DateTimeField(missing=dt.datetime(2000,1,1))
```

This works fine in OO world but when the schema is exported to marshmallow schema, the default value is invalid, as marshmallow's `DateTime` expects a `string`, not a `datetime`.

In this case, a solution is to pass a `string`. This will work in both cases since uMongo's `DateTimeField` can handle a string (by letting `DateTime` deserialize it):

```python
from marshmallow.utils import isoformat

@instance.register
class Whatever(Document):
    date = DateTimeField(missing=isoformat(dt.datetime(2000,1,1)))
```

But maybe the lesson here is that the default values should not be the same in uMongo and Marshmallow schemas.

Following your proposal, we could have this:

- `default`: like you said, used in OO world when data is missing at init / load or from DB
- `missing` : ignored in OO world, used in generated marshmallow schema

Pass both to generated marshmallow schema.

It might sound a a bit awkward, having an attribute that is ignored but used only in the specific as_marshmallow case, but it is actually already the case for any unknown attribute, treated as metadata and passed transparently (description,...). If we implement the logic you're proposing, `missing` will be treated like any other metadata and it will work as I say. Great.

Back to the example, we'd write this

```python
@instance.register
class Whatever(Document):
    date = DateTimeField(
        default=dt.datetime(2000,1,1),
        missing=isoformat(dt.datetime(2000,1,1))
)
```

This looks redundant, though.

We could use `default` for everything and write this:

```python
@instance.register
class Whatever(Document):
    date = DateTimeField(
        default=isoformat(dt.datetime(2000,1,1))
)
```

and pass `default` as both `default` and `missing` in marshmallow schema but I'm not sure it is a good idea to impose a single default value because

- this would prevent making marshmallow schema with different `default`/`missing` values (maybe not a big deal, and after all the schema can always be modified manually)

- there could be cases where a value accepted by marshmallow field is not accepted by overridden `_deserialize` method in corresponding uMongo field.

I don't have any example for the latter and maybe this case could be considered wrong uMongo field design. It should not happen if all uMongo fields are written like this:

```python
class MyField(MyMaField):

    def _deserialize(self, value, attr, data):
        if value is ...:
            return value
        if isinstance(value,...):
            return whatever(value)
        [...]
        return super()._deserialize(value, attr, data)  # Ultimately, call marshmallow field
```

### TL;DR

Currently we're facing an issue for which we have a workaround.

If we do as you propose and don't think too much, there should be no issue anymore...

**Comment by touilleMan** [comment]
I think we should keep thing explicit to avoid misleading use of `missing`/`default`. For example we only use `default` un umongo, raise a runtime error if `missing` is used (to explain how umongo and marshmallow have their behavior changing on this point) and use a `marshmallow_missing`/`marshmallow_default` fields to pass a custom attribute value when using `as_marshmallow_field`.
Of course if `marshamallow_*` is not provided, we fall back on `default` value (serializing it on the fly before marshmallow use it)

I've created a PR which implement this behavior #107, can you have a look and play a bit with it (the subject is pretty complicated, better have two pair of eyes on it !)

**Comment by lafrech** [comment]
We still have the "default value is not validated" issue I described above.

See this test: https://github.com/Scille/umongo/compare/rework-default...Nobatek:rework-default-tests

It is possible to pass a wrong default value. There is no validation and no deserialization.

I might be wrong but I think it would be better to validate/deserialize the default value when loading.

Here is marshmallow behavior:

```python
import datetime as dt

from marshmallow import Schema, fields, validate


class MySchema(Schema):
    name = fields.Str(
        missing='Eric',
        default='Eric',
        validate=validate.OneOf(('Stan', 'Kyle', 'Kenny')))

    date = fields.DateTime(
        missing='1992-06-02T00:00:00',
        default=dt.datetime(1992, 6, 2),
        validate=validate.OneOf((dt.datetime(1992, 6, 2),))
    )


ret = MySchema().load({})
print(ret)
# UnmarshalResult(data={'date': datetime.datetime(1992, 6, 2, 0, 0)}, errors={'name': ['Not a valid choice.']})
# missing is validated/deserialized. an error is raised if the value is wrong.


ret = MySchema().dump({})
print(ret)
# MarshalResult(data={'name': 'Eric', 'date': datetime.datetime(1992, 6, 2, 0, 0)}, errors={})
# default is passed as is without validation/serialization (we should have passed a string value)
```

We should probably validate `default`. At first sight, it is tempting to raise the error at import time, but since `default` can be a callable, it should be done at instantiation time anyway.

**Comment by lafrech** [comment]
This default/validation issue aside, I like the design.

I might have chosen `ma_missing` / `ma_default` to make things shorter, but I don't mind. Anyway, thanks to this it should not happen too often:

> Of course if marshamallow_* is not provided, we fall back on default value (serializing it on the fly before marshmallow use it)

**Comment by lafrech** [comment]
Related Marshmallow issue: https://github.com/marshmallow-code/marshmallow/issues/378

Things might change in Marshmallow 3.

**Comment by touilleMan** [comment]
:'-(
I remember all the suffering for porting umongo when marshmallow 2 arrived... wish this new version won't break too much things...

**Comment by lafrech** [comment]
I've been following the discussions about Marshmallow 3 and from what I gathered, the choices made are likely to make our lives easier.

Among the breaking changes, strict would be True by default or even removed and we'd need to catch `ValidationError`, and load would return the deserialized data rather than the (data, errors) tuple. This will break stuff, but ultimately, the code should be easier to read.

**Comment by lafrech** [comment]
Anyway, my point was not about compatibility. I just meant that if we want to keep as close as possible to Marshmallow's behavior in our missing/default management (which we should, IMHO), we might as well aim at the next version.

Assuming [that proposal](https://github.com/marshmallow-code/marshmallow/issues/378) is implemented, I think we should

- expect deserialized default in uMongo field (like we already do)
- validate default in uMongo field (we currently don't)

Also, when moving to Marshmallow 3, and if the proposal is implemented, we shall not serialize default anymore when using as as marshmallow_*. But currently, we should.

**Comment by lafrech** [comment]
Closed by https://github.com/Scille/umongo/pull/107.

The "default value is not validated"  point still stands, but that's another story. Let's see what happens in Marshmallow 3. Passing defaults in deserialized form raised issues.
